crates/wasi-common/src/fdentry.rs

```diff
@@ -1,35 +1,335 @@
+    }
+
+    fn get_rights_base(&self) -> wasi::__wasi_rights_t {
+        wasi::RIGHTS_REGULAR_FILE_BASE
+    }
+
+    fn get_rights_inheriting(&self) -> wasi::__wasi_rights_t {
+        wasi::RIGHTS_REGULAR_FILE_INHERITING
+    }
+}
+
+/// A clonable read/write directory.
+pub struct VirtualDir {
+    writable: bool,
+    entries: Arc<RefCell<HashMap<PathBuf, Box<dyn VirtualFile>>>>,
+}
+
+impl VirtualDir {
+    pub fn new(writable: bool) -> Self {
+        VirtualDir {
+            writable,
+            entries: Arc::new(RefCell::new(HashMap::new())),
+        }
+    }
+
+    pub fn with_file<P: AsRef<Path>>(self, file: Box<dyn VirtualFile>, path: P) -> Self {
+        self.entries.borrow_mut().insert(path.as_ref().to_owned(), file);
+        self
+    }
+}
+
+impl VirtualFile for VirtualDir {
+    fn try_clone(&self) -> io::Result<Box<dyn VirtualFile>> {
+        Ok(Box::new(
+            VirtualDir {
+                writable: self.writable,
+                entries: Arc::clone(&self.entries)
+            }
+        ))
+    }
+
+    fn readlinkat(&self, path: &Path) -> Result<String> {
+        // no symlink support, so always say it's invalid.
+        Err(Error::EINVAL)
+    }
+
+    fn openat(&self, path: &Path, read: bool, write: bool, oflags: u16, fs_flags: u16) -> Result<Box<dyn VirtualFile>> {
+        let mut entries = self.entries.borrow_mut();
+        match entries.entry(path.to_owned()) {
+            Entry::Occupied(e) => e.get().try_clone().map_err(Into::into),
+            Entry::Vacant(v) => {
+                if self.writable {
+                    println!("created new file: {}", path.display());
+                    v.insert(Box::new(InMemoryFile::new())).try_clone()
+                        .map_err(Into::into)
+                } else {
+                    Err(Error::EACCES)
+                }
+            }
+        }
+    }
+
+    fn create_directory(&self, path: &Path) -> Result<()> {
+        let mut entries = self.entries.borrow_mut();
+        match entries.entry(path.to_owned()) {
+            Entry::Occupied(e) => Err(Error::EEXIST),
+            Entry::Vacant(v) => {
+                if self.writable {
+                    println!("created new virtual directory at: {}", path.display());
+                    v.insert(Box::new(VirtualDir::new(false)));
+                    Ok(())
+                } else {
+                    Err(Error::EACCES)
+                }
+            }
+        }
+    }
+
+    fn write_vectored(&mut self, iovs: &[io::IoSlice]) -> Result<usize> {
+        Err(Error::EBADF)
+    }
+
+    fn get_file_type(&self) -> wasi::__wasi_filetype_t {
+        wasi::__WASI_FILETYPE_DIRECTORY
+    }
+
+    fn get_rights_base(&self) -> wasi::__wasi_rights_t {
+        wasi::RIGHTS_DIRECTORY_BASE
+    }
+
+    fn get_rights_inheriting(&self) -> wasi::__wasi_rights_t {
+        wasi::RIGHTS_DIRECTORY_INHERITING
+    }
+}
+
+pub(crate) enum File {
+    OsFile(OsFile),
```

Comment:
wasi-common has been reorganized a bit since where I branched, so this is a big ol' merge conflict waiting to happen, now.